### PR TITLE
Upgrade NetCDF output writer to write arbitrary outputs for LESbrary use

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR allows `NetCDFOutputWriter` to write arbitrary outputs, similar to how we pass functions that return data to be saved to disk with the JLD2 output writer, but now with metadata! The `dimensions` kwarg allows you to pass a dictionary of dimensions to apply to any outputs.

I also cleaned up the `NetCDFOutputWriter` a bit and changed the name of the time dimension from `"Time"` to `"time"`.

@suyashbire1 Let me know if I did anything wrong.

I added a pretty comprehensive test that shows how you might write a mix of scalars, profiles, and slices to a NetCDF file with grid/unit/name metadata and global attributes.

Here's what the API looks like right now:
```julia
# Define scalar, vector, 2D slice, and 3D field outputs
f(model) = model.clock.time^2
g(model) = @. model.clock.time * exp(model.grid.zC)
h(model) = @. model.clock.time * sin(model.grid.xC) * cos(model.grid.yC')

outputs = Dict("scalar" => f, "profile" => g, "slice" => h)
dims = Dict("scalar" => (), "profile" => ("zC",), "slice" => ("xC", "yC"))

output_attributes = Dict(
    "scalar"  => Dict("longname" => "Some scalar", "units" => "bananas"),
    "profile" => Dict("longname" => "Some vertical profile", "units" => "watermelons"),
    "slice"   => Dict("longname" => "Some slice", "units" => "mushrooms")
)

global_attributes = Dict("location" => "Bay of Fundy", "onions" => 7)

simulation.output_writers[:fruits] =
    NetCDFOutputWriter(
        model, outputs; frequency=1, filename="test_function_outputs.nc", dimensions=dims,
        global_attributes=global_attributes, output_attributes=output_attributes)

run!(simulation)
```

### CODE REVIEW

File: src/OutputWriters/netcdf_output_writer.jl [hunk 1]
```diff
@@ -59,138 +56,144 @@ function write_grid(model; filename="./grid.nc", mode="c",
...
+        for (dim_name, dim_array) in dims
+            defVar(ds, dim_name, dim_array, (dim_name,),
+                   compression=compression, attrib=dim_attribs[dim_name])
         end
     end
 
     return nothing
 end
 
+const default_output_attributes =
```
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl]: might be good to add one for `b` (buoyancy)

Comment (glwagner) [reply src/OutputWriters/netcdf_output_writer.jl]: with units `m / s^2`

Comment (glwagner) [reply src/OutputWriters/netcdf_output_writer.jl]: Also we should clarify that `T` is Conservative Temperature and `S` is absolute salinity. This matters for non-linear equations of state.

File: src/OutputWriters/netcdf_output_writer.jl [hunk 2]
```diff
@@ -59,138 +56,144 @@ function write_grid(model; filename="./grid.nc", mode="c",
...
-    close(ds)
-    return field
-end
+Base.close(ow::NetCDFOutputWriter) = close(ow.dataset)
 
 """
     slice(field; slice_kw...)
 
-For internal use only. Returns a slice for a field based on its dimensions and the supplied slices in `slice_kw`.
+For internal use only. Returns a slice for a field based on its dimensions and the
```
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl]: Do we have to say "For internal use only"?

Comment (ali-ramadhan) [reply src/OutputWriters/netcdf_output_writer.jl]: True. Not exporting it implies it's mainly for internal use.

**Comment by ali-ramadhan** [comment]
The NetCDF tests fail on GPU so gotta look into that.

> Is this documented? Is there an example?

No and no, but yeah we should update model setup documentation and show an example of how to use it. Vorticity example might be good. Might also be nice to have the example save a horizontal average to NetCDF so we can also show how to save outputs with different dimensions.

File: src/OutputWriters/netcdf_output_writer.jl (hunk 2 above)
Comment (glwagner) [reply src/OutputWriters/netcdf_output_writer.jl]: Let the user decide!

**Comment by glwagner** [comment]
> No and no, but yeah we should update model setup documentation and show an example of how to use it. Vorticity example might be good. Might also be nice to have the example save a horizontal average to NetCDF so we can also show how to save outputs with different dimensions.

Maybe just get this merged and open an issue?

File: \ [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    return nothing
+end
+
+const default_output_attributes = Dict(
+    "u" => Dict("longname" => "Velocity in the x-direction", "units" => "m/s"),
+    "v" => Dict("longname" => "Velocity in the y-direction", "units" => "m/s"),
+    "w" => Dict("longname" => "Velocity in the z-direction", "units" => "m/s"),
+    "b" => Dict("longname" => "Buoyancy",                    "units" => "m/s²"),
+    "T" => Dict("longname" => "Conservative temperature",    "units" => "K"),
+    "S" => Dict("longname" => "Absolute aalinity",           "units" => "g/kg")
```
Comment (glwagner) [inline \]: typo

File: src/OutputWriters/netcdf_output_writer.jl [hunk 4]
```diff
@@ -59,138 +56,145 @@ function write_grid(model; filename="./grid.nc", mode="c",
...
     return nothing
 end
 
+const default_output_attributes = Dict(
+    "u" => Dict("longname" => "Velocity in the x-direction", "units" => "m/s"),
+    "v" => Dict("longname" => "Velocity in the y-direction", "units" => "m/s"),
+    "w" => Dict("longname" => "Velocity in the z-direction", "units" => "m/s"),
+    "b" => Dict("longname" => "Buoyancy",                    "units" => "m/s²"),
+    "T" => Dict("longname" => "Conservative temperature",    "units" => "K"),
+    "S" => Dict("longname" => "Absolute aalinity",           "units" => "g/kg")
```
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl]: Typo here. There is also a spurious file called `/` above.

**Comment by glwagner** [comment]
@ali-ramadhan a spurious file should be deleted before this is merged.

**Comment by ali-ramadhan** [comment]
Ah thanks for catching that, no idea how it happened.

**Comment by ali-ramadhan** [comment]
> Maybe just get this merged and open an issue?

Yeah sure, there are a few other things that need updating too.

**Pull request merged**